Sources/NIOChatServer/main.swift

```diff
@@ -177,7 +177,11 @@ let channel = try { () -> Channel in
     }
 }()
 
-print("ChatServer started and listening on \(channel.localAddress!)")
+if let localAddress = channel.localAddress {
+    print("ChatServer started and listening on \(localAddress)")
+} else {
+    print("ChatServer started but the address could not be determined.")
+}
```

Comment:
I don't think we need to change these: we will always know our local address at this stage.

Reply:
@Lukasa I was unsure about that one. Sure, this can't actually happen but then always having a always having a force unwrap here is also not great in the examples... So I thought that is okay and I thought @normanmaurer approved too but he actually didn't, he just ran the tests. Apologies!

Reply:
That was my motivation in making this change.  I wanted to remove the force unwraps in the examples because I know that avoiding force unwraps is a common Swift development technique.

Reply:
IMO, blindly replacing force unwraps with conditional binding is no better than blindly force unwrapping values. A force unwrap should only be used when the value being optional would represent *programmer error*, or when it is literally not possible to recover from the optional case.

Put another way, this is a perfectly good force-unwrap that (I hope) no-one would suggest we should replace with `guard let`:

```swift
let x = URL(string: "https://github.com/")!
```

A more complex example that I consider to be the same is this code block: https://github.com/apple/swift-nio/blob/2839ef0a3cfb157deca38d1a2540a3ec0d55fa5a/Sources/NIOTLS/SNIHandler.swift#L197-L228

This block contains three force-unwraps, but all of them are "safe" because of the `guard` at the top of the function. If any of those force-unwraps ever fails, that failure will be because I have made an error in assuming that `tlsRecordHeaderLength` is at least `5`, or because of an error in the `ByteBuffer` implementation. Throwing an error in that case would be entirely misleading. This is exactly like force-unwrapping the URL above: the force unwrap is justified by my other code, and I simply cannot express to the compiler that I know this will always succeed in any way other than a force unwrap.

Returning to `localAddress`: my argument is that NIO makes a promise that `localAddress` will always be well-defined and non-optional once the `bind` promise returns, until the file descriptor is closed. If that promise is not kept, NIO is quite severely wrong about the state of the world, and it should probably crash rather than keep running in this confused state. For this reason, I consider the force-unwrap here to be totally safe. It's a short way to say "this value being `nil` here indicates a malfunctioning program".

Reply:
@Lukasa agreed if you can _know_ that it can never be `nil` and that is the case for the code you quote.
But we're talking about the example code here and users don't _know_ this can not be `nil` as they're not NIO devs.
I don't really have a good suggestion because we can't change the API (as not _all_ channels have `localAddress`es) but we also shouldn't make everybody force unwrap here I don't think. Maybe the solution is to have a `String` returning function that is non-optional and just fills in `"<not available>"` or something if unavailable.

Reply:
@Lukasa everything you write is totally accurate, yet often the user may just want to print a string and I think many people just don't like to force unwrap for that...
And sure, a better comment would probably help here to give the users confidence on when then can force unwrap this 'safely'.

I don't really know what to do here.

Reply:
@Lukasa and @weissi Thank you for much for the insightful feedback.  After reading through the explanation on specific cases where the local address could not be found, I agree that the, "address could not be determined," message is misleading.  I also agree that a force unwrap failure here signifies a much larger issue in NIO.  I do however feel that more accurate information on why the force unwrap failed would be beneficial as well.  As a middle ground between using a force unwrap and the current patch in place I think the best option would be to use a `guard` statement to unwrap localAddress.  If a failure should ever occur I an error message provides more specific information on what could be happening in these specific failures.

The new PR addressing this feedback is up here: https://github.com/apple/swift-nio/pull/273
Thank you very much for the time and the explanation provided.